pkg/statistics/histogram.go

```diff
@@ -1507,56 +1498,139 @@ func MergePartitionHist2GlobalHist(sc *stmtctx.StatementContext, hists []*Histog
+				}
+				// If buckets[i-1].lower >= currentLeftMost, this bucket is totally inside. So it can be totally merged.
+				if res >= 0 {
+					sum += buckets[i-1].Count
+					mergeBuffer = append(mergeBuffer, buckets[i-1])
+					continue
+				}
+				// Now buckets[i-1].lower < currentLeftMost < buckets[i-1].upper
+				overlapping := calcFraction4Datums(buckets[i-1].lower, buckets[i-1].upper, currentLeftMost)
+				overlappedCount := int64(float64(buckets[i-1].Count) * overlapping)
+				overlappedNDV := int64(float64(buckets[i-1].NDV) * overlapping)
+				sum += overlappedCount
+				buckets[i-1].Count -= overlappedCount
+				buckets[i-1].NDV -= overlappedNDV
+
+				// Cut it.
+				cutBkt := newbucket4MergingForRecycle()
+				buckets[i-1].upper.Copy(cutBkt.upper)
+				currentLeftMost.Copy(cutBkt.lower)
+				cutBkt.Count = overlappedCount
+				cutBkt.NDV = overlappedNDV
+				mergeBuffer = append(mergeBuffer, cutBkt)
+				cutAndFixBuffer = append(cutAndFixBuffer, cutBkt)
+
+				currentLeftMost.Copy(buckets[i-1].upper)
+				buckets[i-1].Repeat = 0
+			}
+			var merged *bucket4Merging
+			if len(cutAndFixBuffer) == 0 {
+				merged, err = mergePartitionBuckets(sc, buckets[i:r])
+				if err != nil {
+					return nil, err
+				}
+			} else {
+				// The content in the merge buffer don't need a re-sort since we just fix some lower bound for them.
+				mergeBuffer = append(mergeBuffer, buckets[leftMostValidPosForNonOverlapping:r]...)
+				merged, err = mergePartitionBuckets(sc, mergeBuffer)
+				if err != nil {
+					return nil, err
+				}
+				for _, bkt := range cutAndFixBuffer {
+					releasebucket4MergingForRecycle(bkt)
+				}
+				// The buckets in buckets[i:origI] needs a re-sort.
+				err = sortBucketsByUpperBound(sc.TypeCtx(), buckets[i:leftMostValidPosForNonOverlapping])
+				if err != nil {
+					return nil, err
+				}
+				// After the operation, the buckets in buckets[i:origI] contains two kinds of buckets:
+				// 1. The buckets that are totally inside the merged bucket. => lower_bound >= currentLeftMost
+				//    It's not changed. [lower_bound_i, upper_bound_i] with lower_bound_i >= currentLeftMost
+				// 2. The buckets that are overlapped with the merged bucket. lower_bound < currentLeftMost < upper_bound
+				//    After cutting, the remained part is [lower_bound_i, currentLeftMost]
+				// To do the next round of merging, we need to kick out the 1st kind of buckets.
+				// And after the re-sort, the 2nd kind of buckets will be in the front.
+				leftMostInvalidPosForNextRound := leftMostValidPosForNonOverlapping
+				for ; leftMostInvalidPosForNextRound > i; leftMostInvalidPosForNextRound-- {
+					res, err := buckets[leftMostInvalidPosForNextRound-1].lower.Compare(sc.TypeCtx(), currentLeftMost, collate.GetBinaryCollator())
+					if err != nil {
+						return nil, err
+					}
+					// Once the lower bound < currentLeftMost, we'll skip all the 1st kind of bucket.
+					// We can break here.
+					if res < 0 {
+						break
```

Comment:
Since this part of logic is very complicated, I suggest we add an assert here to make sure `upper == currentLeftMost` is true for all buckets in `[i,leftMostInvalidPosForNextRound-1]`.

Reply:
good idea, added.